packages/scoped-elements/src/ScopedElementsMixin.js

```diff
@@ -1,194 +1,105 @@
-        throw new Error('The `scopeName` option is required.');
-      }
-      const { scopeName, eventContext } = options;
-
-      const templateCache = getTemplateCache(eventContext);
-      const tagsCache = getTagsCache(eventContext);
-      const { scopedElements } = this;
-
-      return super.render(template, container, {
-        ...options,
-        templateFactory: scopedElementsTemplateFactory(
-          scopeName,
-          scopedElements,
-          templateCache,
-          tagsCache,
-        ),
-      });
+    /** @type {ShadowRootInit} */
+    static get shadowRootOptions() {
+      return this.__shadowRootOptions;
     }
 
-    /**
-     * Defines a scoped element
-     *
-     * @param {string} tagName
-     * @param {typeof HTMLElement} klass
-     */
-    defineScopedElement(tagName, klass) {
-      return defineScopedElement(tagName, klass, getTagsCache(this));
+    /** @type {ShadowRootInit} */
+    static set shadowRootOptions(value) {
+      this.__shadowRootOptions = value;
+    }
+
+    /** @type {CSSResultFlatArray} */
+    static get elementStyles() {
+      return this.__elementStyles;
+    }
+
+    static set elementStyles(styles) {
+      this.__elementStyles = styles;
+    }
+
+    // either TS or ESLint will complain here
+    // eslint-disable-next-line no-unused-vars
+    constructor(..._args) {
+      super();
+      /** @type {RenderOptions} */
+      this.renderOptions = this.renderOptions || undefined;
     }
 
     /**
-     * Returns a scoped tag name
+     * Obtains the CustomElementRegistry if specified.
      *
-     * @deprecated Please, use the instance method instead of the static one. This static method is not able to
-     * obtain the tagName of lazy defined elements, while the instance one is.
-     * @param {string} tagName
-     * @returns {string|undefined}
+     * @return {CustomElementRegistry}
      */
-    static getScopedTagName(tagName) {
-      // @ts-ignore
-      const klass = this.scopedElements[tagName];
+    get registry() {
+      if (!this._registry) {
+        this._registry = new CustomElementRegistry();
+      }
+
+      return this._registry;
+    }
+
+    createRenderRoot() {
+      const {
+        scopedElements,
+        shadowRootOptions,
+        elementStyles,
+      } = /** @type {typeof ScopedElementsHost} */ (this.constructor);
+
+      Object.entries(scopedElements).forEach(([tagName, klass]) =>
+        this.registry.define(tagName, klass),
+      );
 
-      return klass
-        ? registerElement(tagName, klass, getTagsCache(this))
-        : getTagsCache(this).get(tagName);
+      /** @type {ShadowRootInit} */
+      const options = {
+        mode: 'open',
+        ...shadowRootOptions,
+        customElements: this.registry,
+      };
+
+      this.renderOptions.creationScope = this.attachShadow(options);
+
+      if (this.renderOptions.creationScope instanceof ShadowRoot)
+        adoptStyles(this.renderOptions.creationScope, elementStyles);
```

Comment:
why do we need this logic in the scoped elements mixin?

Reply:
because adoptStyles only takes a ShadowRoot

Reply:
In the previous version of lit adoptStyles was called outside createShadowRoot, but I  the new version is called inside it. If we are overriding it we should call adoptStyles also